packages/@eventual/core-runtime/src/handlers/entity-stream-worker.ts

```diff
@@ -45,20 +44,34 @@ export function createEntityStreamWorker(
 
         return { failedItemIds: result?.failedItemIds ?? [] };
       } else {
+        const itemsByKey: Record<string, EntityStreamItem<any>[]> = {};
+        items.forEach((item) => {
+          const normalizedKey = normalizeCompositeKey(entity, item.key);
+          const serializedKey = JSON.stringify(normalizedKey);
+          (itemsByKey[serializedKey] ??= []).push(item);
+        });
+
         const results = await promiseAllSettledPartitioned(
-          items,
-          async (item) => {
-            return await streamHandler.handler(item, context);
+          Object.entries(itemsByKey),
+          async ([, itemGroup]) => {
+            for (const i in itemGroup) {
+              const item = itemGroup[i]!;
+              try {
+                const result = await streamHandler.handler(item, context);
+                // if the handler doesn't fail and doesn't return false, continue
+                if (result !== false) {
+                  continue;
+                }
+              } catch {}
+              // if the handler fails or returns false, return the rest of the items
+              return itemGroup.slice(Number(i)).map((i) => i.id);
```

Comment:
This seems unconventional

Reply:
I don't like for-loops, continue, or empty catches either, but it seemed more important to normalize the false and error case into one line and handle the success case explicitly.

1. sequential promises
2. return all future events that should also be marked as failed to retry.